pkg/protocol/protocol_fork.go

```diff
@@ -82,57 +83,96 @@ func (p *Protocol) onForkDetected(fork *chainmanager.Fork) {
-	candidateEngine, err := p.engineManager.ForkEngineAtSlot(snapshotTargetIndex)
+	candidateEngineInstance, err := p.engineManager.ForkEngineAtSlot(snapshotTargetIndex)
 	if err != nil {
 		p.ErrorHandler()(ierrors.Wrap(err, "error creating new candidate engine"))
 		return
 	}
 
 	// Set the chain to the correct forking point
-	candidateEngine.SetChainID(fork.ForkingPoint.ID())
+	candidateEngineInstance.SetChainID(fork.ForkingPoint.ID())
 
 	// Attach the engine block requests to the protocol and detach as soon as we switch to that engine
-	wp := candidateEngine.Workers.CreatePool("CandidateBlockRequester", 2)
-	detachRequestBlocks := candidateEngine.Events.BlockRequester.Tick.Hook(func(blockID iotago.BlockID) {
+	detachRequestBlocks := candidateEngineInstance.Events.BlockRequester.Tick.Hook(func(blockID iotago.BlockID) {
 		p.networkProtocol.RequestBlock(blockID)
-	}, event.WithWorkerPool(wp)).Unhook
+	}, event.WithWorkerPool(candidateEngineInstance.Workers.CreatePool("CandidateBlockRequester", 2))).Unhook
+
+	var detachProcessCommitment, detachMainEngineSwitched func()
+
+	cleanupFunc := func() {
+		detachRequestBlocks()
+		detachProcessCommitment()
+		detachMainEngineSwitched()
+
+		candidateEngineInstance.Shutdown()
+		if err := candidateEngineInstance.RemoveFromFilesystem(); err != nil {
+			p.ErrorHandler()(ierrors.Wrapf(err, "error cleaning up candidate engine %s from file system", candidateEngineInstance.Name()))
+		}
+	}
 
 	// Attach slot commitments to the chain manager and detach as soon as we switch to that engine
-	detachProcessCommitment := candidateEngine.Events.Notarization.SlotCommitted.Hook(func(details *notarization.SlotCommittedDetails) {
+	detachProcessCommitment = candidateEngineInstance.Events.Notarization.SlotCommitted.Hook(func(details *notarization.SlotCommittedDetails) {
+		// Check whether the commitment produced by syncing the candidate engine is actually part of the forked chain.
+		if fork.ForkedChain.LatestCommitment().ID().Index() >= details.Commitment.Index() {
+			forkedChainCommitmentID := fork.ForkedChain.Commitment(details.Commitment.Index()).ID()
+			if forkedChainCommitmentID != details.Commitment.ID() {
+				cleanupFunc()
+				return
+			}
+		}
+
 		p.ChainManager.ProcessCandidateCommitment(details.Commitment)
-	}, event.WithWorkerPool(candidateEngine.Workers.CreatePool("ProcessCandidateCommitment", 2))).Unhook
+	}, event.WithWorkerPool(candidateEngineInstance.Workers.CreatePool("ProcessCandidateCommitment", 2))).Unhook
 
-	p.Events.MainEngineSwitched.Hook(func(_ *engine.Engine) {
+	// Clean up events when we switch to the candidate engine.
+	detachMainEngineSwitched = p.Events.MainEngineSwitched.Hook(func(_ *engine.Engine) {
 		detachRequestBlocks()
 		detachProcessCommitment()
-	}, event.WithMaxTriggerCount(1))
+	}, event.WithMaxTriggerCount(1)).Unhook
+
+	// Clean up candidate engine if we never switch to it.
+	go func() {
+		timer := time.NewTimer(10 * time.Minute)
```

Comment:
I never see this timer getting cancelled if we switch to the candidate engine properly